sdk/tracing/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java

```diff
@@ -132,180 +160,72 @@ public boolean isEndRequired() {
-        // avoid blocking the producer thread.
-        ArrayList<ReadableSpan> spansCopy;
-        synchronized (monitor) {
-          // If still maxExportBatchSize elements in the queue better to execute an extra
-          if (spansList.size() < maxExportBatchSize) {
-            do {
-              // In the case of a spurious wakeup we export only if we have at least one span in
-              // the batch. It is acceptable because batching is a best effort mechanism here.
-              try {
-                monitor.wait(scheduleDelayMillis);
-              } catch (InterruptedException ie) {
-                // Preserve the interruption status as per guidance and stop doing any work.
-                Thread.currentThread().interrupt();
-                return;
-              }
-            } while (spansList.isEmpty());
+    final CompletableResultCode result = spanExporter.export(forExport);
+    result.whenComplete(
+        new Runnable() {
+          @Override
+          public void run() {
+            if (!result.isSuccess()) {
+              logger.log(Level.FINE, "Exporter failed");
+            }
+            if (queue.size() >= maxExportBatchSize) {
+              exportBatch();
+            } else {
+              scheduleNextExport();
+            }
           }
-          spansCopy = new ArrayList<>(spansList);
-          spansList.clear();
-        }
-        // Execute the batch export outside the synchronized to not block all producers.
-        exportBatches(spansCopy);
-      }
-    }
-
-    private void shutdown() {
-      forceFlush();
-      timer.cancel();
-      spanExporter.shutdown();
-    }
-
-    private void forceFlush() {
-      ArrayList<ReadableSpan> spansCopy;
-      synchronized (monitor) {
-        spansCopy = new ArrayList<>(spansList);
-        spansList.clear();
-      }
-      // Execute the batch export outside the synchronized to not block all producers.
-      exportBatches(spansCopy);
-    }
-
-    private void exportBatches(ArrayList<ReadableSpan> spanList) {
-      // TODO: Record a counter for pushed spans.
-      for (int i = 0; i < spanList.size(); ) {
-        int lastIndexToTake = Math.min(i + maxExportBatchSize, spanList.size());
-        onBatchExport(createSpanDataForExport(spanList, i, lastIndexToTake));
-        i = lastIndexToTake;
-      }
-    }
-
-    private static List<SpanData> createSpanDataForExport(
-        List<ReadableSpan> spanList, int startIndex, int endIndex) {
-      List<SpanData> spanDataBuffer = new ArrayList<>(endIndex - startIndex);
-      for (int i = startIndex; i < endIndex; i++) {
-        spanDataBuffer.add(spanList.get(i).toSpanData());
-        // Remove the reference to the ReadableSpan to allow GC to free the memory.
-        spanList.set(i, null);
-      }
-      return Collections.unmodifiableList(spanDataBuffer);
-    }
+        });
+
+    // Timeout if result doesn't complete soon enough.
```

Comment:
Not specific to this PR: Why this timeout is the configuration of BSP and not of the exporter?